gateway-js/src/index.ts

```diff
@@ -1227,40 +1052,33 @@ export class ApolloGateway implements GraphQLService {
         }
         return;
       case 'loaded':
-        this.state = { phase: 'stopped' }; // nothing to do (we're not polling)
-        return;
-      case 'waiting to poll': {
-        // If we're waiting to poll, we can synchronously transition to fully stopped.
-        // We will terminate the current pollServices call and it will succeed quickly.
-        const doneWaiting = this.state.doneWaiting;
-        clearTimeout(this.state.pollWaitTimer);
-        this.state = { phase: 'stopped' };
-        doneWaiting();
-        return;
-      }
-      case 'polling': {
-        // We're in the middle of running updateSchema. We need to go into 'stopping'
-        // mode and let this run complete. First we set things up so that any concurrent
-        // calls to stop() will wait until we let them finish. (Those concurrent calls shouldn't
-        // just wait on pollingDonePromise themselves because we want to make sure we fully
-        // transition to state='stopped' before the other call returns.)
-        const pollingDonePromise = this.state.pollingDonePromise;
-        let stoppingDone: () => void;
+        const stoppingDonePromise = this.performCleanup();
         this.state = {
           phase: 'stopping',
-          stoppingDonePromise: new Promise<void>((res) => {
-            stoppingDone = res;
-          }),
+          stoppingDonePromise,
         };
-        await pollingDonePromise;
+        await stoppingDonePromise;
         this.state = { phase: 'stopped' };
-        stoppingDone!();
         return;
+      case 'updating schema': {
+        // This should never happen since schema updates are sync and `stop`
```

Comment:
This can't happen for *concurrency* reasons but can happen for *reentrancy* reasons (if stop is explicitly called from the schema change notification callback). Dunno why anyone would do this and it's fine to throw an error but probably should be more like "`ApolloGateway.stop` should not be called from inside a schema change listener" (and we can probably make it work later if it's actually useful somehow)

Reply:
7b7101d7